localstack-core/localstack/services/sns/v2/provider.py

```diff
@@ -118,6 +164,360 @@ def set_topic_attributes(
+        # TODO: validate format on the token (seems to be 288 hex chars)
+        # this request can come from any http client, it might not be signed (we would need to implement
+        # `authenticate_on_unsubscribe` to force a signing client to do this request.
+        # so, the region and account_id might not be in the request. Use the ones from the topic_arn
+        try:
+            parsed_arn = parse_arn(topic_arn)
+        except InvalidArnException:
+            raise InvalidParameterException("Invalid parameter: Topic")
+
+        store = self.get_store(account_id=parsed_arn["account"], region=parsed_arn["region"])
+
+        # it seems SNS is able to know what the region of the topic should be, even though a wrong topic is accepted
+        if parsed_arn["region"] != get_region_from_subscription_token(token):
+            raise InvalidParameterException("Invalid parameter: Topic")
+
+        subscription_arn = store.subscription_tokens.get(token)
+        if not subscription_arn:
+            raise InvalidParameterException("Invalid parameter: Token")
+
+        subscription = store.subscriptions.get(subscription_arn)
+        if not subscription:
+            # subscription could have been deleted in the meantime
+            raise InvalidParameterException("Invalid parameter: Token")
+
+        # ConfirmSubscription is idempotent
+        if subscription.get("PendingConfirmation") == "false":
+            return ConfirmSubscriptionResponse(SubscriptionArn=subscription_arn)
+
+        subscription["PendingConfirmation"] = "false"
+        subscription["ConfirmationWasAuthenticated"] = "true"
+
+        return ConfirmSubscriptionResponse(SubscriptionArn=subscription_arn)
+
+    def list_subscriptions(
+        self, context: RequestContext, next_token: nextToken = None, **kwargs
+    ) -> ListSubscriptionsResponse:
+        store = self.get_store(context.account_id, context.region)
+        subscriptions = [
+            select_from_typed_dict(Subscription, sub) for sub in list(store.subscriptions.values())
+        ]
+        paginated_subscriptions = PaginatedList(subscriptions)
+        page, next_token = paginated_subscriptions.get_page(
+            token_generator=lambda x: get_next_page_token_from_arn(x["SubscriptionArn"]),
+            page_size=100,
+            next_token=next_token,
+        )
+
+        response = ListSubscriptionsResponse(Subscriptions=page)
+        if next_token:
+            response["NextToken"] = next_token
+        return response
+
+    def list_subscriptions_by_topic(
+        self, context: RequestContext, topic_arn: topicARN, next_token: nextToken = None, **kwargs
+    ) -> ListSubscriptionsByTopicResponse:
+        self._get_topic(topic_arn, context)
+        parsed_topic_arn = parse_and_validate_topic_arn(topic_arn)
+        store = self.get_store(parsed_topic_arn["account"], parsed_topic_arn["region"])
+        sns_subscriptions = store.get_topic_subscriptions(topic_arn)
```

Comment:
this could be simplified by getting the topic and accessing the topic's `subscriptions`:


```suggestion
        # TODO: fix `_get_topic` to consider properly use the region and allow cross region access
        topic = self._get_topic(topic_arn, context)
        sns_subscriptions = topic["subscriptions"]
```